pkg/rpc/auth_test.go

```diff
@@ -274,6 +274,72 @@ func TestAuthenticateTenant(t *testing.T) {
 	}
 }
 
+func BenchmarkAuthenticate(b *testing.B) {
+	correctOU := []string{security.TenantsOU}
+	stid := roachpb.SystemTenantID
+	for _, tc := range []struct {
+		name         string
+		systemID     roachpb.TenantID
+		ous          []string
+		commonName   string
+		rootDNString string
+		nodeDNString string
+	}{
+		{name: "tenTen", systemID: stid, ous: correctOU, commonName: "10"},
+		{name: "rootDN", systemID: stid, ous: nil, commonName: "foo", rootDNString: "CN=foo"},
+		{name: "nodeDN", systemID: stid, ous: nil, commonName: "foo", nodeDNString: "CN=foo"},
+		{name: "commonRoot", systemID: stid, ous: nil, commonName: "root"},
+		{name: "commonNode", systemID: stid, ous: nil, commonName: "node"},
+	} {
+		b.Run(tc.name, func(b *testing.B) {
+			var err error
+			if tc.rootDNString != "" {
+				err = security.SetRootSubject(tc.rootDNString)
+				if err != nil {
+					b.Fatalf("could not set root subject DN, err: %v", err)
+				}
+			}
+			if tc.nodeDNString != "" {
+				err = security.SetNodeSubject(tc.nodeDNString)
+				if err != nil {
+					b.Fatalf("could not set node subject DN, err: %v", err)
+				}
+			}
+			defer func() {
+				security.UnsetRootSubject()
+				security.UnsetNodeSubject()
+			}()
+
+			cert := &x509.Certificate{
+				Subject: pkix.Name{
+					CommonName:         tc.commonName,
+					OrganizationalUnit: tc.ous,
+				},
+			}
+			cert.RawSubject, err = asn1.Marshal(cert.Subject.ToRDNSequence())
+			if err != nil {
+				b.Fatalf("unable to marshal rdn sequence to raw subject, err: %v", err)
+			}
+			tlsInfo := credentials.TLSInfo{
+				State: tls.ConnectionState{
+					PeerCertificates: []*x509.Certificate{cert},
+				},
+			}
+			p := peer.Peer{AuthInfo: tlsInfo}
+			ctx := peer.NewContext(context.Background(), &p)
+
+			b.ResetTimer()
+			var clusterSettings map[settings.InternalKey]settings.EncodedValue
+			for i := 0; i < b.N; i++ {
+				_, err := rpc.TestingAuthenticateTenant(ctx, tc.systemID, clusterSettings)
```

Comment:
I checked out your PR to look at the flamegraph for the remaining three allocs -  `TestingAuthenticateTenant` is not a good entry point for this benchmark as it makes a `settings.NewUpdater` etc, which dominates the allocations:


```
./dev bench --ignore-cache --stream-output --bench-mem ./pkg/rpc --filter Authenticate/commonRoot --test-args '-test.memprofile=mem.pb.gz -test.benchtime=50000x -test.cpu=8' && pprof -http :6060 mem.pb.gz
```

<img width="1728" alt="image" src="https://github.com/user-attachments/assets/aedb335b-f426-460c-b9bd-e1019ad3002e">

Could you refactor this entry point to take a `settings.Values` instead, or make a new one?

I hacked this using this diff

and the result shows that we're actually allocation-free on the `commonRoot` and `commonName` benchmarks.

```
BenchmarkAuthenticate/tenTen-8             50000                43.08 ns/op            0 B/op          0 allocs/op
BenchmarkAuthenticate/rootDN-8             50000              1165 ns/op             464 B/op         25 allocs/op
BenchmarkAuthenticate/nodeDN-8             50000              1149 ns/op             464 B/op         25 allocs/op
BenchmarkAuthenticate/commonRoot-8         50000                80.63 ns/op            0 B/op          0 allocs/op
BenchmarkAuthenticate/commonNode-8         50000                80.48 ns/op            0 B/op          0 allocs/op
```

Reply:
For completeness here's the alloc_objects for the combined `BenchmarkAuthenticate` with the spurious allocs suppressed
<img width="1722" alt="image" src="https://github.com/user-attachments/assets/9a065abb-2963-490a-a536-0501a4ae1141">

Reply:
Good idea. I was lazy and just stopped tinkering with the benchmark when I observed the reduction in allocs. The benchmark is more accurate now.

I'll look into the `CheckCertDNMatchesRootDNorNodeDN` allocations and see if they can be reduced easily. I have zero experience here, so it'll take me some time. I also don't want to spend too many cycles on it since it's not coming up in our target workload.

Reply:
Yep, if it's not in the target let's not spend more time on it now.
<!-- Reviewable comment -OBziVruDvxhf9p9gORt:bri31if -->
<!-- Sent from Reviewable.io -->